### PR TITLE
GELF (Graylog2) logging backend

### PR DESCRIPTION
**Pull request by irq0**
Add graylog2 as logger target. Format log entries as GELF [0] and send them via UDP to log server

New config options
 -  log_to_graylog
 -  err_to_graylog
 -  log_graylog_host
 -  log_graylog_port

[0] https://www.graylog2.org/resources/gelf

Signed-off-by: Marcel Lauhoff <ml@irq0.org>

### CODE REVIEW

**Comment by irq0** [comment]
I also wrote a blog article about it: http://irq0.org/articles/ceph/graylog2

File: src/log/Log.cc [hunk 1]
```diff
@@ -141,6 +153,32 @@ void Log::set_stderr_level(int log, int crash)
...
+{
+  m_fsid = fsid;
+}
+
+void Log::set_graylog_destination(string host, int port)
+{
+  boost::asio::ip::udp::resolver resolver(m_graylog_io_service);
+  boost::asio::ip::udp::resolver::query query(host,
+					      boost::lexical_cast<string>(port));
+  m_graylog_endpoint = *resolver.resolve(query);
```
Comment (liewegas) [inline src/log/Log.cc]: i would expect some error checking here?

File: src/log/Log.cc [hunk 2]
```diff
@@ -239,8 +278,52 @@ void Log::_flush(EntryQueue *t, EntryQueue *requeue, bool crash)
...
+
+	try {
+	  boost::asio::ip::udp::socket socket(m_graylog_io_service);
+	  socket.open(m_graylog_endpoint.protocol());
+	  socket.send_to(boost::asio::buffer(zos.str()), m_graylog_endpoint);
+	} catch (boost::system::system_error const& e) {
+	  /* The above code fails until the configuration is set */
+	}
+      }
+    }
```
Comment (liewegas) [inline src/log/Log.cc]: This looks functionally correct, but probably pretty slow relative to the other backends.  Formatter could be a static member of Log and re-used, along with fsid_str.  Maybe the zlib stream?  Not sure.

Anyway, that all can come later.

File: configure.ac [hunk 3]
```diff
@@ -755,6 +755,10 @@ AC_CHECK_LIB(boost_system-mt, main, [],
     [AC_CHECK_LIB(boost_system, main, [],
         AC_MSG_NOTICE(["Boost system library not found."]))])
 
+AC_CHECK_LIB(boost_iostreams-mt, main, [],
+    [AC_CHECK_LIB(boost_iostreams, main, [],
+        AC_MSG_NOTICE(["Boost iostreams library not found."]))])
```
Comment (liewegas) [inline configure.ac]: Is this a new package?  Ideally iostreams-mt is already part of the boost libraries in build-depends in debian/control and ceph.spec.in ....

**Comment by liewegas** [comment]
Looks good!  A somewhat related item would be to make the cluster log (mon/LogMonitor.cc) also target graylog.

There is also a new audit log mechanism that was just added that has some key/value data in it, but it's not being formatted or structured in a generic (JSON) way yet... if there is something simple we can/should do there before releasing hammer that gets it into a better form for consumption by the likes of logstash it would be great to do it now...

**Comment by irq0** [comment]
I did a bit of refactoring: The GELF/Graylog related code now lives in the Ceph::Log::Graylog class. I also added more error checking and improved resource reuse.

Some thoughts about thread safety:
Graylog::log_entry is not thread safe, since it uses lots of object resources without locking. But, as far as I understand the locking in the Log class, it should be safe as _flush serializes calls.

I haven't had a close look at mon/LogMonitor.cc yet, but I thing the refactored Graylog class fit into there somehow. I'll look into it.


Audit log: I'm not familiar with it, but for logstash, graylog, etc consumption GELF [42] is quite practical. It's JSON; compressed; key/value. The only gotcha is that non well-defined fields must start with '_'.

[42] https://www.graylog2.org/resources/gelf

**Comment by liewegas** [comment]
@irq0: did you see my question about the new boost dependency?  If this doesn't cause build issues we could merge it even without cluster log support, although it would definitely be preferable if that were there too (so that there's parity with file vs syslog vs graylog).

**Comment by magicrobotmonkey** [comment]
What happened to this?

File: configure.ac (hunk 3 above)
Comment (irq0) [inline configure.ac]: boost_iostreams-mt is new. 
I added it to debian/control. ceph.spec lists "boost-devel" as build dependency, which includes iostreams.

**Comment by irq0** [comment]
Sorry, the branch fell victim to my cleanup script :(

**Comment by irq0** [comment]
Reopened as #4774. github won't let me reopen this one

**Pull request closed without merging**